### PR TITLE
Dual Replication - Primary changes

### PR DESCRIPTION
**Pull request by shourya035**
### Description
[Describe what this change achieves]

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -5199,6 +5209,10 @@ public AsyncIOProcessor<Translog.Location> getTranslogSyncProcessor() {
         return translogSyncProcessor;
     }
 
+    public boolean ongoingEngineMigration() {
+        return RemoteStoreUtils.isMigrationDirectionSet(clusterService) == true;
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java]: This will return true for all the shards during migration . Will that have any side effects ?

File: server/src/main/java/org/opensearch/index/shard/CheckpointRefreshListener.java [hunk 2]
```diff
@@ -43,7 +43,14 @@ protected boolean performAfterRefreshWithPermit(boolean didRefresh) {
...
             && shard.getReplicationTracker().isPrimaryMode()
-            && !shard.indexSettings.isSegRepWithRemoteEnabled()) {
+            && !shard.indexSettings.isSegRepWithRemoteEnabled()
+            /*
+             During remote store migration, the isSegRepWithRemoteEnabled criteria would return false
+             since we do not alter the remote store based index settings at that stage. Explicitly
+             blocking checkpoint publication from this refresh listener since it ends up interfering
+             with the RemoteStoreRefreshListener invocation
+             */
+            && !shard.ongoingEngineMigration()) {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/shard/CheckpointRefreshListener.java]: can we have `isSegRepLocalEnabled` instead of negative checks ?

File: server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java [hunk 3]
```diff
@@ -1248,7 +1254,8 @@ private void createReplicationLagTimers() {
                 if (cps.inSync
                     && replicationGroup.getUnavailableInSyncShards().contains(allocationId) == false
                     && isPrimaryRelocation(allocationId) == false
-                    && latestReplicationCheckpoint.isAheadOf(cps.visibleReplicationCheckpoint)) {
+                    && latestReplicationCheckpoint.isAheadOf(cps.visibleReplicationCheckpoint)
+                    && (ongoingEngineMigration && routingTable.getByAllocationId(allocationId).isAssignedToRemoteStoreNode() == true)) {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java]: why do we need `ongoingEngineMigration` here ?

File: server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java [hunk 4]
```diff
@@ -1096,4 +1098,14 @@ enum IndexRemovalReason {
             REOPENED,
         }
     }
+
+    private void updateShardRoutingWithNode(IndexShardRoutingTable indexShardRoutingTable, DiscoveryNodes nodes) {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java:1106]: what about `createShard` flow ?

File: server/src/main/java/org/opensearch/cluster/routing/ShardRouting.java [hunk 5]
```diff
@@ -75,6 +75,7 @@ public class ShardRouting implements Writeable, ToXContentObject {
     private final long expectedShardSize;
     @Nullable
     private final ShardRouting targetRelocatingShard;
+    private boolean assignedToRemoteStoreNode;
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/ShardRouting.java]: this is not populated by master and only data node . Is it intentional ? If yes, should we make it `Boolean` instead ? Also it is final .

File: server/src/main/java/org/opensearch/action/support/replication/ReplicationModeAwareProxy.java [hunk 6]
```diff
@@ -60,16 +61,21 @@ protected void performOnReplicaProxy(
 
     @Override
     ReplicationMode determineReplicationMode(ShardRouting shardRouting, ShardRouting primaryRouting) {
-
         // If the current routing is the primary, then it does not need to be replicated
         if (shardRouting.isSameAllocation(primaryRouting)) {
             return ReplicationMode.NO_REPLICATION;
         }
-
+        // Perform full replication during primary failover
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/action/support/replication/ReplicationModeAwareProxy.java]: primary relocation.

File: server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java (hunk 4 above)
Comment (shourya035) [reply server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java:1106]: Addressing that

File: server/src/main/java/org/opensearch/cluster/routing/ShardRouting.java (hunk 5 above)
Comment (shourya035) [reply server/src/main/java/org/opensearch/cluster/routing/ShardRouting.java]: Ideally we don't want to serialize the value for this field since this won't be used anywhere else. I have changed this to `Boolean` instead of the primitive type, but I have assigned a default value to prevent any possible NPEs when the shardRouting entry sent from the cluster manager nodes are serialized on the data node.

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java (hunk 1 above)
Comment (shourya035) [reply server/src/main/java/org/opensearch/index/shard/IndexShard.java]: Removed this altogether from `IndexShard`

File: server/src/main/java/org/opensearch/index/shard/CheckpointRefreshListener.java (hunk 2 above)
Comment (shourya035) [reply server/src/main/java/org/opensearch/index/shard/CheckpointRefreshListener.java]: Addressed this. Changed the `isSegRepLocalEnabled` check to ensure that the Checkpoint listeners won't be registered if the node is non-remote

File: server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java (hunk 3 above)
Comment (shourya035) [reply server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java]: Removed this. Relying on the shardRouting and `isSegRepLocalEnabled` check to ensure that the timers are only created if:

- The shard copy is hosted on a remote enabled node
- The index setting has segrep enabled but remote disabled (vanilla segrep)

File: server/src/main/java/org/opensearch/action/support/replication/TransportReplicationAction.java [hunk 7]
```diff
@@ -642,7 +642,7 @@ public void handleException(TransportException exp) {
                         primaryRequest.getPrimaryTerm(),
                         initialRetryBackoffBound,
                         retryTimeout,
-                        indexShard.isRemoteTranslogEnabled()
+                        indexShard.isRemoteTranslogEnabled() || indexShard.indexSettings().isRemoteNode()
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/action/support/replication/TransportReplicationAction.java:645]: Lets subsume the check of `indexShard.indexSettings().isRemoteNode()` inside the `indexShard.isRemoteTranslogEnabled()` check itself?

File: server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java [hunk 8]
```diff
@@ -1089,10 +1089,6 @@ private ReplicationGroup calculateReplicationGroup() {
             newVersion = replicationGroup.getVersion() + 1;
         }
 
-        assert indexSettings().isRemoteTranslogStoreEnabled()
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java:1092]: Lets not remove this assertion. Removing this can pose risks to bugs.

We can check first if index is remote enabled, or if there is an inflight migration, or all tracked shards are replicated.

File: server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java [hunk 9]
```diff
@@ -1248,7 +1244,14 @@ private void createReplicationLagTimers() {
...
                     && isPrimaryRelocation(allocationId) == false
-                    && latestReplicationCheckpoint.isAheadOf(cps.visibleReplicationCheckpoint)) {
+                    && latestReplicationCheckpoint.isAheadOf(cps.visibleReplicationCheckpoint)
+                    /*
+                     Handle remote store migration cases. Replication Lag timers would be created if the node on which primary is hosted is either:
+                     - Segrep enabled without remote store
+                     - Destination replica shard is hosted on a remote store enabled node (Remote store enabled nodes have segrep enabled implicitly)
+                    */
+                    && (indexSettings.isSegRepLocalEnabled() == true
+                        || routingTable.getByAllocationId(allocationId).isAssignedToRemoteStoreNode() == true)) {
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java:1254]: Not really clear on why is this being added?

File: server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java [hunk 10]
```diff
@@ -1511,12 +1520,18 @@ public synchronized void updateFromClusterManager(
...
+        String primaryTargetAllocationId,
+        boolean assignedToRemoteStoreNode
+    ) {
+        /*
+         - If remote translog is enabled, then returns replication mode checking current allocation id against the primary and primary target allocation id.
+         - If remote translog is enabled, then returns true if given allocation id matches the primary or it's relocation target allocation primary and primary target allocation id.
+         - During an ongoing remote migration, the above-mentioned checks are considered when the shard is assigned to a remote store backed node
+         */
+        if (indexSettings().isRemoteTranslogStoreEnabled() || assignedToRemoteStoreNode == true) {
             return (allocationId.equals(primaryAllocationId) || allocationId.equals(primaryTargetAllocationId));
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java:1535]: can we return false if the if condition holds true since the remote translogs would anyways be downloaded on the relocating primary?

File: server/src/main/java/org/opensearch/index/seqno/RetentionLeaseBackgroundSyncAction.java [hunk 11]
```diff
@@ -198,6 +203,20 @@ protected void shardOperationOnReplica(Request request, IndexShard replica, Acti
...
+    public ReplicationMode getReplicationMode(IndexShard indexShard) {
+        /*
+        Unblock PRRL publication on both docrep and remote shard copies
+        during remote store migration. This is done deliberately to ensure
+        data consistency on remote to docrep shard copy failover during the
+        migration process.
+         */
+        if (RemoteStoreUtils.isMigrationDirectionSet(clusterService)) {
+            return ReplicationMode.FULL_REPLICATION;
+        }
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/seqno/RetentionLeaseBackgroundSyncAction.java:216]: Why do we need it on shards that are on remote store already?

Comment (ashking94) [reply server/src/main/java/org/opensearch/index/seqno/RetentionLeaseBackgroundSyncAction.java:216]: Same question applicable for other actions as well.

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java [hunk 12]
```diff
@@ -3328,7 +3333,7 @@ public void syncRetentionLeases() {
                 )
             );
         } else {
-            logger.trace("background syncing retention leases [{}] after expiration check", retentionLeases.v2());
+            logger.info("background syncing retention leases [{}] after expiration check", retentionLeases.v2());
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java]: do we need this with info level?

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java [hunk 13]
```diff
@@ -4621,6 +4627,10 @@ public final boolean isSearchIdleSupported() {
         if (isRemoteTranslogEnabled() || indexSettings.isRemoteNode()) {
             return false;
         }
+        // Explicitly disable search idle during remote store migration
+        if (RemoteStoreUtils.isMigrationDirectionSet(clusterService)) {
+            return false;
+        }
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java]: Curious on why is this required?

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java [hunk 14]
```diff
@@ -340,7 +340,7 @@ private void onSuccessfulSegmentsSync(
         resetBackOffDelayIterator();
         // Set the minimum sequence number for keeping translog
         indexShard.getEngine().translogManager().setMinSeqNoToKeep(lastRefreshedCheckpoint + 1);
-        // Publishing the new checkpoint which is used for remote store + segrep indexes
+        // Publishing the new checkpoint which is used for remote store + segrep indexess
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:343]: lets remove this unintended change?

File: server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java [hunk 15]
```diff
@@ -669,6 +669,11 @@ private void createShard(DiscoveryNodes nodes, RoutingTable routingTable, ShardR
         try {
             final long primaryTerm = state.metadata().index(shardRouting.index()).primaryTerm(shardRouting.id());
             logger.debug("{} creating shard with primary term [{}]", shardRouting.shardId(), primaryTerm);
+            DiscoveryNode targetNode = nodes.getLocalNode();
+            // Set remote attributes on Shard routing if the target node for the shards has remote attributes
+            if (targetNode.isRemoteStoreNode()) {
+                shardRouting.setAssignedToRemoteStoreNode(true);
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java:675]: appears to be error prone, is there a way to have this as final and set inside ctor?

File: server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java [hunk 16]
```diff
@@ -1096,4 +1102,14 @@ enum IndexRemovalReason {
             REOPENED,
         }
     }
+
+    private void updateShardRoutingWithNode(IndexShardRoutingTable indexShardRoutingTable, DiscoveryNodes nodes) {
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java:1106]: is there a way this can be done within ctor only?

**Review by ashking94 — CHANGES_REQUESTED** [review]
Left some comments, please take a look at them.
